pkg/localmetrics/localmetrics.go

```diff
@@ -270,3 +282,71 @@ func (c *MetricsCollector) AddAccountReuseCleanupFailure() {
 func (c *MetricsCollector) SetReconcileDuration(controller string, duration float64) {
 	c.reconcileDuration.WithLabelValues(controller).Observe(duration)
 }
+
+func (c *MetricsCollector) AddAPICall(controller string, req *http.Request, resp *http.Response) {
+	c.apiCallCount.With(prometheus.Labels{
+		"controller": controller,
+		"method":     req.Method,
+		"resource":   resourceFrom(req.URL),
+		"status":     resp.Status,
+	}).Inc()
+}
+
+// resourceFrom normalizes an API request URL, including removing individual namespace and
+// resource names, to yield a string of the form:
+//     $group/$version/$kind[/{NAME}[/...]]
+// or
+//     $group/$version/namespaces/{NAMESPACE}/$kind[/{NAME}[/...]]
+// ...where $foo is variable, {FOO} is actually {FOO}, and [foo] is optional.
+// This is so we can use it as a dimension for the apiCallCount metric, without ending up
+// with separate labels for each {namespace x name}.
+func resourceFrom(url *neturl.URL) (resource string) {
+	defer func() {
+		// If we can't parse, return a general bucket. This includes paths that don't start with
+		// /api or /apis, or paths with fewer than the expected number of components (e.g.
+		// discovery URLs).
+		if r := recover(); r != nil {
+			// TODO(efried): Should we be logging these? I guess if we start to see a lot of them...
+			resource = "{OTHER}"
+		}
+	}()
+
+	tokens := strings.Split(url.Path[1:], "/")
+
+	// First normalize to $group/$version/...
+	switch tokens[0] {
+	case "api":
+		// Core resources: /api/$version/...
+		// => core/$version/...
+		tokens[0] = "core"
+	case "apis":
+		// Extensions: /apis/$group/$version/...
+		// => $group/$version/...
+		tokens = tokens[1:]
+	default:
+		// Something else. Punt.
+		panic(1)
```

Comment:
This will be a problem when we instrument AWS API calls no?

Reply:
When we do that, we'll have to beef up this method to generate a usable resource string. I haven't fully formulated what that should look like, but it for sure doesn't need to have the same format as the kubeclient calls. We can make the prefix `aws/$service/` (e.g. `aws/ec2/`) and go from there. If we ever need to generate graphs for just kube or just AWS, prom lets you use regexes for label matching. Will still need to figure out how to scrub unique names/IDs out.

(BTW, anticipating the need to get the AWS service name is why I made the parameter to this func the whole URL rather than just the Path part as hive did -- IIUC the service name is in the domain, not the path.)

Reply:
Yeah we are not instrumenting the AWS API calls now but good to know you thought of it for when the time comes. And  yes for AWS the service is in the domain. ec2.<region>.amazonaws.com exceptions are global services like route53 where there is no region but we can deal with that when the time comes.